TensorLib/NumpyRepr.lean

```diff
@@ -0,0 +1,278 @@
+  | .ok (x, s) => .ok (some x, s)
+
+def ignore (p : PState T) : PState Unit := do
+  let _ <- tryParse p
+  return ()
+
+-- Does some heavy lifting, parsing both normal ids (map keys) and weird
+-- values like the dtypes
+def parseToken : PState String := do
+  whitespace
+  let s ← get
+  let mut token := ""
+  for i in [s.index : s.headerEnd] do
+    let b := s.source.get! i
+    let c := Char.ofUInt8 b
+    if c.isAlphanum || c = '_' || c = '<' || c = '|' then token := token.push c
+    else break
+  if token.length != 0 then
+    set { s with index := s.index + token.length }
+    return token
+  else
+    .error "Can't parse token"
+
+def consume (c: Char) : PState Unit := do
+  whitespace
+  let s <- get
+  let i := s.index
+  if s.source[i]! == c.toUInt8 then
+    set { s with index := i + 1 }
+  else
+    .error s!"Can't consume {c}. So far we have {repr s}"
+
+def quote : PState Unit := consume '\''
+def colon : PState Unit := consume ':'
+def comma : PState Unit := consume ','
+
+def quoted (p : PState T) : PState T := do
+  quote
+  let x <- p
+  quote
+  return x
+
+partial def parseCommaListAux (p : PState T) (acc : List T) : PState (List T) := do
+  let v <- tryParse p
+  match v with
+  | none => return acc.reverse
+  | some x =>
+    ignore comma
+    parseCommaListAux p (x :: acc)
+
+-- We need to parse both () for shapes and {} for the outer metadata record
+def parseCommaList {T : Type} (start : Char) (end_ : Char) (p : PState T) : PState (List T) := do
+  consume start
+  let xs <- parseCommaListAux p []
+  consume end_
+  return xs
+
+def parseShape : PState Shape := do
+  let xs <- parseCommaList '(' ')' parseToken
+  return xs.map (fun x => x.toNat!)
+
+-- major/minor/header-length
+def parseHeader : PState (Nat × Nat) := do
+  let s <- get
+  let b := s.source
+  if s.index != 0 then .error "Illegal start index"
+  if b.size < 10 then .error "Buffer too small"
+  if b[0]! != 0x93 then .error s!"Invalid first byte: {b[0]!}"
+  if b[1]! != 'N'.toUInt8 then .error "Invalid second byte"
+  if b[2]! != 'U'.toUInt8 then .error "Invalid third byte"
+  if b[3]! != 'M'.toUInt8 then .error "Invalid fourth byte"
+  if b[4]! != 'P'.toUInt8 then .error "Invalid fifth byte"
+  if b[5]! != 'Y'.toUInt8 then .error "Invalid sixth byte"
+  let major := b[6]!.toNat
+  let minor := b[7]!.toNat
+  -- Header length is 2 bytes, little-endian
+  let headerLength := b[8]!.toNat + b[9]!.toNat * 256
+  set { s with index := 10, headerEnd := 10 + headerLength }
+  return (major, minor)
+
+-- There are 3 fields, descr, fortran_order, and shape. We are not supposed to depend
+-- on the ordering of the fields
+def parseOneMetadata : PState Unit := do
+  let id <- quoted parseToken
+  colon
+  if id == "descr" then
+    let v <- quoted parseToken
+    let d <- NumpyDtype.fromString v
+    modify (fun s => { s with descr := some d })
+  else if id == "fortran_order" then
+    let v <- parseToken
+    let s <- get
+    match v with
```

Comment:
This is where I though you may use the bool fromString?

Reply:
Yep.